File: Algorithm/MinimumTimeDifference.cpp
class Solution {
public:
    int findMinDifference(vector<string>& timePoints) {
        constexpr int kMins = 24 * 60;
        set<int> seen;
        for (const string& time : timePoints) {
            int m = stoi(time.substr(0, 2)) * 60 + stoi(time.substr(3));
            if (!seen.insert(m).second) return 0;
        }
        
        int ans = (*seen.begin() - *seen.rbegin() + kMins) % kMins;
        const int* l = nullptr;
        for (const int& t : seen) {
            if (l) ans = min(ans, t - *l);
            l = &t;
        }
        return ans;
    }
};

File: Algorithm/MinimumTimeDifference.java
// Solution:
// Time complexity: O(nlog1440)
// Space complexity: O(n)
class Solution {
    public int findMinDifference(List<String> timePoints) {
        final int kMins = 24 * 60;
        TreeSet<Integer> seen = new TreeSet<>();
        for (final String time : timePoints) {
            int m = Integer.valueOf(time.substring(0, 2)) * 60 + Integer.valueOf(time.substring(3));
            if (!seen.add(m)) return 0;
        }
        
        int ans = (seen.first() - seen.last() + kMins) % kMins;
        int l = 0;
        for (final Integer t : seen) {
            if (l != 0) ans = Math.min(ans, t - l);
            l = t;
        }
        return ans;
    }
}
